internal/chart/helm_client.go

```diff
@@ -105,15 +144,55 @@ func (c HelmClient) UpdateChart(tv TemplateValuer, config ChartConfig, opts ...I
 		namespace: c.namespace,
 		cli:       c.c,
 	}
+	shouldUpdate, err := c.isHelmChartStatusActionable(c.statusFunc, appName, helmStatusActionMapUpdate)
+	if err != nil || !shouldUpdate {
+		return nil, err
+	}
 	return updateClient.Run(appName, chrt, vals)
 }
 
 // DeleteChart uninstalls the app's helm release. It doesn't return an error if the release is not found.
 func (c HelmClient) DeleteChart(appName string) error {
+	shouldDelete, err := c.isHelmChartStatusActionable(c.statusFunc, appName, helmStatusActionMapDelete)
+	if err != nil || !shouldDelete {
+		return err
+	}
 	uninstall := action.NewUninstall(c.cfg)
-	_, err := uninstall.Run(appName)
+	_, err = uninstall.Run(appName)
 	if err != nil && errors.Is(err, driver.ErrReleaseNotFound) {
 		return nil
 	}
 	return err
 }
+
+// getHelmStatus returns the Status, and error for an app
+func getHelmStatus(cfg *action.Configuration, appName string) (release.Status, error) {
+	statusClient := action.NewStatus(cfg)
+	status, err := statusClient.Run(appName)
+	if err != nil {
+		if errors.Is(err, driver.ErrReleaseNotFound) || status.Info == nil {
+			return notFound, nil
+		}
+		return "", err
+	}
+	return status.Info.Status, nil
+}
+
+// isHelmChartStatusActionable returns true if the statusFunc returns an actionable status according to the statusActionMap, false if the status is
+// non-actionable (e.g. "not-found" status for a "delete" action), and an error if the status requires a wait-retry. The retry is expected to be
+// executed by the calling reconciler's inherent looping.
+func (c HelmClient) isHelmChartStatusActionable(statusFunc statusFunc, appName string, statusActionMap map[release.Status]int) (bool, error) {
+	status, err := statusFunc(c.cfg, appName)
+	if err != nil {
+		return false, err
+	}
+	switch statusActionMap[status] {
+	case noAction:
+		c.log.Info(fmt.Sprintf("helm chart for app %s release already in state %s - no action required", appName, status))
+		return false, nil
+	case takeAction:
+		return true, nil
+	default:
```

Comment:
would it help if we implement something like https://github.com/theketchio/ketch/blob/b9e8b85febdc84adba0e61bda88c1ba2994a261b/internal/chart/helm_client.go#L129-L136

i

Reply:
Yes, that makes sense. This PR's addition almost duplicates that `FirstDeployed.Before` check, but considers additional `statuses`. I removed the original block of code.